### PR TITLE
PIP-1007-mappability

### PR DESCRIPTION
**Pull request by ottojolanki**

### CODE REVIEW

File: wdl/tasks/awk.wdl [hunk 1]
```diff
@@ -62,3 +62,50 @@ task shift_bed_reads_start_and_end_range {
...
+
+task clean_reference_fasta_headers {
+    input {
+        File fasta
+        Resources resources
+        String out = "cleaned.fa" 
+    }
+
+    command <<<
+        awk '/^>/{$0=$1} 1' ~{fasta} \
```
Comment (keenangraham) [inline wdl/tasks/awk.wdl:74]: Looks good but can we mutliline these like https://github.com/ENCODE-DCC/dnase-seq-pipeline/blob/ca6bfbb3172cc490be7fc9bf10bd99ca129579f3/wdl/tasks/awk.wdl#L45

File: wdl/tasks/hotspot1.wdl [hunk 2]
```diff
@@ -37,3 +38,41 @@ task runhotspot {
...
     }
 }
+
+
+task enumerate_uniquely_mappable_space {
+    input {
+        BowtieIndex bowtieindex
+        File cleaned_fasta
+        Int kmer_length
+        String out = "enumerated_space.bed"
```
Comment (keenangraham) [inline wdl/tasks/hotspot1.wdl:48]: Should come after Resources.

File: wdl/tasks/sed.wdl [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+task remove_whitespace_from_end_of_lines{
+    input {
+        File input_file
+        Resources resources
+        String out = "out.txt"
+    }
+
+    command {
+        sed 's/[[:blank:]]*$//' ~{input_file} \
```
Comment (keenangraham) [inline wdl/tasks/sed.wdl:15]: Multiline?

File: wdl/subworkflows/build_mappable_only_bed.wdl [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import "../tasks/sed.wdl"
+
+
+workflow build_mappable_only_bed {
+    input {
+        BedopsSortBedParams sortbedparams
+        BowtieIndex bowtieindex
+        File reference_genome
+        Resources resources
+        Int kmer_length
```
Comment (keenangraham) [inline wdl/subworkflows/build_mappable_only_bed.wdl:16]: Should be above Resources.

File: wdl/subworkflows/build_mappable_only_bed.wdl [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+    
+    String mappable_output = basename(reference_genome, ".fa") + ".K" + kmer_length + ".mappable_only.bed"
```
Comment (keenangraham) [inline wdl/subworkflows/build_mappable_only_bed.wdl:19]: Lol.

File: wdl/subworkflows/build_mappable_only_bed.wdl [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        input:
+            bowtieindex=bowtieindex,
+            cleaned_fasta=clean_reference_fasta_headers.cleaned_fasta,
+            kmer_length=kmer_length,
+            resources=resources,
+    }
+
+    call awk.merge_adjacent_bed {
+        input:
+            bed=enumerate_uniquely_mappable_space.enumerated_space,
```
Comment (keenangraham) [inline wdl/subworkflows/build_mappable_only_bed.wdl:37]: enumerated_space -> enumerated_space_bed?

File: wdl/tasks/hotspot1.wdl [hunk 7]
```diff
@@ -37,3 +38,41 @@ task runhotspot {
...
+        ln ~{bowtieindex.rev_ebwt2} .
+        perl $(which enumerateUniquelyMappableSpace.pl) \
+            ~{kmer_length} \
+            ~{bowtieindex_prefix} \
+            ~{cleaned_fasta} \
+            > ~{out}
+    }
+
+    output {
+        File enumerated_space = out
```
Comment (keenangraham) [inline wdl/tasks/hotspot1.wdl:69]: enumerated_space_bed?

File: wdl/subworkflows/build_mappable_only_bed.wdl [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+import "../tasks/awk.wdl"
+import "../tasks/bedops.wdl"
+import "../tasks/hotspot1.wdl"
+import "../tasks/sed.wdl"
+
+
+workflow build_mappable_only_bed {
+    input {
+        BedopsSortBedParams sortbedparams
```
Comment (keenangraham) [inline wdl/subworkflows/build_mappable_only_bed.wdl:12]: sortbedparams -> params

File: wdl/subworkflows/build_mappable_only_bed.wdl [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    call sed.remove_whitespace_from_end_of_lines {
+        input:
+            input_file=merge_adjacent_bed.adjacent_merged_bed,
+            resources=resources,
+    }
+
+    call bedops.sort_bed {
+        input:
+            out = mappable_output,
+            params=sortbedparams,
```
Comment (keenangraham) [inline wdl/subworkflows/build_mappable_only_bed.wdl:50]: params=params or params={}

File: wdl/subworkflows/build_mappable_only_bed.wdl [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    call sed.remove_whitespace_from_end_of_lines {
+        input:
+            input_file=merge_adjacent_bed.adjacent_merged_bed,
+            resources=resources,
+    }
+
+    call bedops.sort_bed {
+        input:
+            out = mappable_output,
```
Comment (keenangraham) [inline wdl/subworkflows/build_mappable_only_bed.wdl:49]: Remove space between =.

File: tests/integration/json/test_build_mappable_only_bed.json [hunk 11]
```diff
@@ -0,0 +1,18 @@
+{
+    "test_build_mappable_only_bed.build_mappable_only_bed.resources": {
+          "cpu": 1,
+          "memory_gb": 1,
+          "disks": "local-disk 10 SSD"
+    },
+    "test_build_mappable_only_bed.build_mappable_only_bed.reference_genome": "tests/data/ref/chr22.fa",
+    "test_build_mappable_only_bed.build_mappable_only_bed.sortbedparams": {},
```
Comment (keenangraham) [inline tests/integration/json/test_build_mappable_only_bed.json:8]: Remove. We should bind this in the actual subworkflow.

File: wdl/subworkflows/build_mappable_only_bed.wdl (hunk 8 above)
Comment (keenangraham) [reply wdl/subworkflows/build_mappable_only_bed.wdl:12]: We should bind params here, not in input JSON. Something like `BedopsSortBedParams params = {}`

See e.g.: https://github.com/ENCODE-DCC/dnase-seq-pipeline/blob/98a5792859d0f92427d4c37ddc46328a8ebe9f1c/wdl/subworkflows/get_first_read_in_pair_from_bam.wdl#L12

File: wdl/tasks/awk.wdl (hunk 1 above)
Comment (ottojolanki) [reply wdl/tasks/awk.wdl:74]: sure.

File: wdl/tasks/awk.wdl [hunk 12]
```diff
@@ -95,7 +97,10 @@ task merge_adjacent_bed {
     }
 
     command {
-        awk -f $(which merge_adjacent_bed.awk) ~{bed} \
+        awk \
+            -f \
```
Comment (keenangraham) [inline wdl/tasks/awk.wdl:101]: I think -f and the file go together on same line.

File: tests/unit/wdl/test_hotspot1_enumerate_uniquely_mappable_space.wdl [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+        Int kmer_length
+        Resources resources
+    }
+
+    call hotspot1.enumerate_uniquely_mappable_space {
+        input:
+            bowtieindex=bowtieindex,
+            cleaned_fasta=cleaned_fasta,
+            kmer_length=kmer_length,
+            resources=resources
```
Comment (keenangraham) [inline tests/unit/wdl/test_hotspot1_enumerate_uniquely_mappable_space.wdl:20]: Trailing comma.

File: wdl/tasks/awk.wdl [hunk 14]
```diff
@@ -62,3 +62,53 @@ task shift_bed_reads_start_and_end_range {
...
+
+task merge_adjacent_bed {
+    input {
+        File bed
+        Resources resources
+        String out = "adjacent_merged.bed"
+    }
+
+    command {
+        awk -f $(which merge_adjacent_bed.awk) \
```
Comment (keenangraham) [inline wdl/tasks/awk.wdl:100]: Hmm.
```
awk \
    -f $(which merge_adjacent_bed.awk) \
    ~{bed} \
    > ~{out}
```

File: tests/unit/wdl/test_hotspot1_enumerate_uniquely_mappable_space.wdl (hunk 13 above)
Comment (keenangraham) [inline tests/unit/wdl/test_hotspot1_enumerate_uniquely_mappable_space.wdl:20]: Needs trailing comma.

File: wdl/tasks/sed.wdl [hunk 15]
```diff
@@ -0,0 +1,30 @@
+version 1.0
+
+
+import "../structs/resources.wdl"
+
+
+task remove_whitespace_from_end_of_lines{
```
Comment (keenangraham) [inline wdl/tasks/sed.wdl:7]: You don't have to change this but maybe `remove_trailing_whitespaces`?

File: wdl/tasks/sed.wdl [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+    command {
+        sed \
+            's/[[:blank:]]*$//' \
+            ~{input_file} \
+            > ~{out}
+    }
+
+    output {
+        File end_whitespace_trimmed = out
```
Comment (keenangraham) [inline wdl/tasks/sed.wdl:22]: `trailing_whitespace_trimmed`

File: wdl/tasks/hotspot1.wdl [hunk 17]
```diff
@@ -37,3 +38,41 @@ task runhotspot {
         disks: resources.disks
     }
 }
+
+
+task enumerate_uniquely_mappable_space {
+    input {
+        BowtieIndex bowtieindex
```
Comment (keenangraham) [inline wdl/tasks/hotspot1.wdl:45]: `bowtie_index` to be consistent with https://github.com/ENCODE-DCC/dnase-seq-pipeline/blob/98a5792859d0f92427d4c37ddc46328a8ebe9f1c/wdl/subworkflows/build_bowtie_index.wdl#L20

File: wdl/tasks/hotspot1.wdl [hunk 18]
```diff
@@ -37,3 +38,41 @@ task runhotspot {
...
+        File cleaned_fasta
+        Int kmer_length
+        Resources resources
+        String out = "enumerated_space.bed" 
+    }
+
+    String bowtieindex_prefix = basename(bowtieindex.ebwt_1, ".1.ebwt")
```
Comment (keenangraham) [inline wdl/tasks/hotspot1.wdl:52]: `bowtie_index_prefix`

File: wdl/tasks/hotspot1.wdl [hunk 19]
```diff
@@ -37,3 +38,41 @@ task runhotspot {
...
+    command {
+        ln ~{bowtieindex.ebwt_1} .
+        ln ~{bowtieindex.ebwt_2} .
+        ln ~{bowtieindex.ebwt_3} .
+        ln ~{bowtieindex.ebwt_4} .
+        ln ~{bowtieindex.rev_ebwt1} .
+        ln ~{bowtieindex.rev_ebwt2} .
+        perl $(which enumerateUniquelyMappableSpace.pl) \
+            ~{kmer_length} \
+            ~{bowtieindex_prefix} \
```
Comment (keenangraham) [inline wdl/tasks/hotspot1.wdl:63]: `bowtie_index_prefix`

File: tests/unit/wdl/test_hotspot1_enumerate_uniquely_mappable_space.wdl (hunk 13 above)
Comment (ottojolanki) [reply tests/unit/wdl/test_hotspot1_enumerate_uniquely_mappable_space.wdl:20]: oh man, we need a functional wdl linter.

File: wdl/tasks/awk.wdl (hunk 14 above)
Comment (ottojolanki) [reply wdl/tasks/awk.wdl:100]: going thru all the combinations possible.

File: wdl/tasks/sed.wdl (hunk 15 above)
Comment (ottojolanki) [reply wdl/tasks/sed.wdl:7]: I like that naming better than the original. Just being a clumsy foreigner here..

**Pull request merged**